### PR TITLE
Metaball rendering example

### PR DESCRIPTION
**Pull request by ataber**
I'm attempting to copy [this cool Three.js demo](https://www.clicktorelease.com/code/bumpy-metaballs/) as much as possible, both for my own learning purposes and for people to be able to compare the two implementations. Outstanding issues: 

- The linked demo has an effect where the rim of the balls look transparent, and I haven't implemented that yet. 

---
This change is [Reviewable](https://reviewable.io/reviews/mikolalysenko/regl/287)

### CODE REVIEW

**Comment by Erkaman** [comment]
I shall take a look at the demo, and see if I can figure out the issue.

**Comment by Erkaman** [comment]
Alright, I examined the performance issues, and here's what I found:

* When managing dynamic geometry with lots of vertices, you are better off managing your own vertex buffers. This is slightly more inconvenient, but it's not that bad. But right now, you are allocating new buffers every single frame(that's how it is implemented internally), and for big buffers, this results in bad performance. Anyways, you need to make sure that you preallocate the vertex buffers in the beginning, and then you update your buffers through `subdata`.  I've made a [gist](https://gist.github.com/Erkaman/bf8c6c7345a76f422a56fffa52212e7b) that shows how to do this. See also the [cloth simulation example](https://github.com/mikolalysenko/regl/blob/gh-pages/example/cloth.js) for another example

* The other problem is not your fault actually. It seems like @mikolalysenko sort of overestimated his implementation of `isosurface`. The first problem is that every time you generate an isosurface, all the arrays needed to contain the data of the isosurface are all allocated from scratch, in the current implementation. But since those arrays are huge, and because you are generating the isosurface every frame, you get lots of garbage collection and terrible performance. Instead, what you should do is that you should preallocate  those arrays in the beginning, and then reuse them throughout the script. This is what the demo you linked to is doing, as you can see [here](https://github.com/mrdoob/three.js/blob/dev/examples/js/MarchingCubes.js#L40-L60)


However, the current API of isosurface doesn't support reusing the arrays, so it can't currently be done. Later, I will try and extend the API and make a pull-request to [isosurface](https://github.com/mikolalysenko/isosurface), and fix this(but you can also give it a stab, if you feel like it).

* Finally, you are using `angle-normals` to compute the normals. But this means that you have to iterate over the isosurface geometry data twice, instead of once. But you can actually compute the normals as part of the isosurface creation by estimating the gradients of the implicit function, so one pass is all you should need. So that is another way of optimizing the implementation. But this is yet another thing that `isosurface` doesn't support. So we'll have to again modify `isosurface` to fix that.

**Comment by ataber** [comment]
Another thing that came up while profiling the code is that the `field` function is taking ~4secs of CPU time. This seems like the candidate for why even the initial render is very slow. I'm comparing my implementation of metaballs with the one [here](https://threejsdoc.appspot.com/doc/three.js/src.source/extras/objects/MarchingCubes.js.html) that is used for the Three.JS demo, and I don't see where the huge difference in performance comes from. Their code basically computes each metaball's contribution separately and then adds that contribution to an array. I'm using closures instead, and then adding up the contribution of each metaball at each point (x,y,z). I'm not sure why there should be a huge difference in performance because of that? Perhaps there's another crucial difference I'm not seeing

**Comment by ataber** [comment]
Ugh. Using a `for` loop instead of `map` and then `reduce` decreases the time spent in `field` by a factor of 10. I don't understand why :/

**Comment by mikolalysenko** [comment]
Instead of isosurface you should use surface nets:

https://github.com/mikolalysenko/surface-nets

Isosurface is not optimized.

**Comment by Erkaman** [comment]
Even with `surface-nets`, you still have to do two passes over the data, since it doesn't compute the normals for you. And extending  `surface-nets` so that it generates normals seems quite difficult. 

So I'd still recommend you to use and extend `isosurface`, I think.

**Comment by Erkaman** [comment]
I don't know either why that is faster. Perhaps using `map` and `reduce` prevents the compiler from doing certain optimizations.  But if it is faster, then you should probably do that instead of `map` and stuff.

File: example/metaball.js [hunk 1]
```diff
@@ -119,21 +121,92 @@ const drawMetaballs = regl({
...
+  fill(fieldArray, (i, j, k) => {
+    let val = 0
+
+    for (let n = 0; n < numblobs; n++) {
+      let ball = balls[n]
+      let zs = ball.center[2] * size
+      let ys = ball.center[1] * size
+      let xs = ball.center[0] * size
+
+      var min_z = Math.floor(zs - r)
```
Comment (ataber) [inline example/metaball.js]: I really dislike this code style, but I'm trying to mirror as much as possible the example code. When it looks the same as [this](https://www.clicktorelease.com/code/bumpy-metaballs/) then I'll refactor it to make it more elegant code.

File: example/metaball.js [hunk 2]
```diff
@@ -142,66 +136,51 @@ const position = (time, i) => {
...
-  let balls = Array(numblobs).fill().map((_, i) => {
-    let bx, by, bz
-    [bx, by, bz] = position(time, i)
-    return {
-      functional: metaball(bx, by, bz, strength, subtract),
-      center: [bx, by, bz]
-    }
-  })
-
-  fill(fieldArray, (i, j, k) => {
```
Comment (ataber) [inline example/metaball.js]: @mikolalysenko for some reason when I prepare `fieldArray` with data before initializing an `ndarray` it is far more performant than using `fill`. I don't understand why, do you have insight on this? The implementation of `fill` is pretty bare-bones, so I don't see where the performance hit comes from. Is it perhaps related to the inherent cost to call a function in JS?

**Comment by ataber** [comment]
So `surface-nets` is running way faster than `isosurface` but not just because it's a more optimized library. Because it takes in an `ndarray` instead of a function, I can populate the array with a `for` loop (which apparently is much more performant than calling a function).

Anyways, this is ready for a code review! Thanks for your help

**Comment by Erkaman** [comment]
Alright, time for code review then.

**Comment by Erkaman** [comment]
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[package.json, line 49 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRutj9rLB9chvRFjdH:-KQRutj9rLB9chvRFjdI:b-oezio4) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/package.json#L49)):*
> ```JSON
>     "ncp": "^2.0.0",
>     "ndarray": "^1.0.18",
>     "ndarray-fill": "^1.0.1",
> ```

Wait, where are you using `ndarray-fill`?

---

*[example/metaball.js, line 169 \[r2\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KPu5bed8YCQWJkebDjm-r2-169:-KQRwNp26BlZku3vUWkD:b-5re2ji) ([raw file](https://github.com/regl-project/regl/blob/adfd9128bb872703a9e94d9919f9100f7832d115/example/metaball.js#L169)):*
<details><summary><i>Previously, ataber wrote…</i></summary>
> I really dislike this code style, but I'm trying to mirror as much as possible the example code. When it looks the same as [this](https://www.clicktorelease.com/code/bumpy-metaballs/) then I'll refactor it to make it more elegant code.
</details>
Well, maybe you should refactor it now then? You can probably make it a bit shorter by using some kind of clamp() function.

---

*[example/metaball.js, line 133 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRv5oMGuGwo4shhcab:-KQRv5oMGuGwo4shhcac:b4qxgg3) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L133)):*
> ```JavaScript
>       vec3 bump3 = texture2D(normalMap, coord3).rgb;
> 
>       vec3 blended_bump = bump1 * blend_weights.xxx +
> ```

You're doing triplanar texturing here, and this technique is not entirely obvious the first time you see it.
So you should add a comment that says that you are using triplanar texturing, for improving the readability, I think. 

---

*[example/metaball.js, line 152 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRyJZovKqQUdzdOKuR:-KQRyJZovKqQUdzdOKuS:b896mwk) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L152)):*
> ```JavaScript
> 
>       vec3 r = reflect(normalize(vU), normalize(finalNormal));
>       float m = 2.0 * sqrt(r.x * r.x + r.y * r.y + (r.z + 1.0) * (r.z + 1.0));
> ```

Explain that you are using a matcap here. And maybe link to this article as well:
https://www.clicktorelease.com/blog/creating-spherical-environment-mapping-shader
Because this is not entirely obvious the first time you see it. This would make it easier to read. 

And for introducing more structure to the shader, maybe you should do the matcap stuff in a function of it's own? And also the triplanar texturing in a function of its own? Because the shader is slightly messy now. It's just one big block of code, basically. 

---

*[example/metaball.js, line 162 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRvfw8V_JdmdKxs3CX:-KQRvfw9k3etBYpmxjOl:b-j85se6) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L162)):*
> ```JavaScript
>       base = vec3(1.) - (vec3(1.) - base) * (vec3(1.) - base);
> 
>       float nn = .05 * random(vec3(1.), length(gl_FragCoord));
> ```

You're doing some dithering here I think? Some better variable names would make that a bit clearer, I think. And maybe also a comment. 

---

*[example/metaball.js, line 260 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRwoHIU2QTUD18hluG:-KQRwoHJKTjah70ssxNr:b97mwoq) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L260)):*
> ```JavaScript
>     sphereTexture: {
>       type: 'image',
>       src: 'spheretexture.jpg',
> ```

Should be changed to `assets/spheretexture.jpg`, so that we can easily run the example as `budo metaball.js` from the `example` folder.  That's how we are doing with all the other example.s 

---

*[example/metaball.js, line 263 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRxClTq9j3OTaGjNJP:-KQRxClTq9j3OTaGjNJQ:b-9f70zs) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L263)):*
> ```JavaScript
>       parser: (data) => regl.texture({
>         data: data,
>         wrapT: 'clamp',
> ```

Slightly redundant. Because as you can see in the docs, both wrapT and wrapS default to 'clamp'
https://github.com/regl-project/regl/blob/gh-pages/API.md#textures

---

*[example/metaball.js, line 271 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRx39GgOmk6lTZ010t:-KQRx39GgOmk6lTZ010u:b-emabei) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L271)):*
> ```JavaScript
>     normalTexture: {
>       type: 'image',
>       src: 'normaltexture.jpg',
> ```

Also fix this path.

---

*[example/assets/normaltexture.jpg, line 0 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRtonFcBjFeAOESut1:-KQRtonFcBjFeAOESut2:b-fjr405) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/assets/normaltexture.jpg#L0)):*
What's the license of this texture? If it is creative commons or public domain, it is fine, but we need to know where it came from, so that we can ask the original author for permission(in case it is not creative commons).

---

*[example/assets/spheretexture.jpg, line 0 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRuI8L7FKfSZgDJLjQ:-KQRuI8L7FKfSZgDJLjR:b8ue3af) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/assets/spheretexture.jpg#L0)):*
Also for this one, we need to know the license.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/287)*

**Comment by Erkaman** [comment]
Alright, did some code review on reviewable now.

**Comment by ataber** [comment]
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[example/assets/normaltexture.jpg, line  \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRtonFcBjFeAOESut1:-KQSUNFWyMh463mFFBKA:bhtm5wp) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/assets/normaltexture.jpg#L)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> What's the license of this texture? If it is creative commons or public domain, it is fine, but we need to know where it came from, so that we can ask the original author for permission(in case it is not creative commons).
</details>
Unfortunately I have no idea what the license is, and a google reverse image search only shows other people using the texture on Pinterest, github projects, etc, no original source. Is there a standard place to find free textures like this?

---

*[example/assets/spheretexture.jpg, line  \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRuI8L7FKfSZgDJLjQ:-KQSUneOq7R9yjdl68OU:bsw8ep7) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/assets/spheretexture.jpg#L)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Also for this one, we need to know the license.
</details>
Same as the other texture.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/287)*

**Comment by Erkaman** [comment]
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[package.json, line 49 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRutj9rLB9chvRFjdH:-KQUJPzBzlh06nWN9Nim:b2fbt1l) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/package.json#L49)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Wait, where are you using `ndarray-fill`?
</details>
Alright, problem solved then. 

---

*[example/metaball.js, line 169 \[r2\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KPu5bed8YCQWJkebDjm-r2-169:-KQUK3rrZ9LRUw0jzZh3:b53yryj) ([raw file](https://github.com/regl-project/regl/blob/adfd9128bb872703a9e94d9919f9100f7832d115/example/metaball.js#L169)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Well, maybe you should refactor it now then? You can probably make it a bit shorter by using some kind of clamp() function.
</details>
Problem solved

---

*[example/metaball.js, line 133 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRv5oMGuGwo4shhcab:-KQUJlc1gPwvrqwF9AzU:b-iygngd) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L133)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> You're doing triplanar texturing here, and this technique is not entirely obvious the first time you see it.
> So you should add a comment that says that you are using triplanar texturing, for improving the readability, I think. 
</details>
Alright problem solved then

---

*[example/metaball.js, line 162 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRvfw8V_JdmdKxs3CX:-KQUJutH8wPMkEX3Cpdm:bxxl90q) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L162)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> You're doing some dithering here I think? Some better variable names would make that a bit clearer, I think. And maybe also a comment. 
</details>
Problen solved

---

*[example/metaball.js, line 260 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRwoHIU2QTUD18hluG:-KQUKK4ZbxRPsIpeIHIg:b53yryj) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L260)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Should be changed to `assets/spheretexture.jpg`, so that we can easily run the example as `budo metaball.js` from the `example` folder.  That's how we are doing with all the other example.s 
</details>
Problem solved

---

*[example/metaball.js, line 263 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRxClTq9j3OTaGjNJP:-KQUKMuWm7RYmPpoBW_D:b53yryj) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L263)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Slightly redundant. Because as you can see in the docs, both wrapT and wrapS default to 'clamp'
> https://github.com/regl-project/regl/blob/gh-pages/API.md#textures
</details>
Problem solved

---

*[example/metaball.js, line 271 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRx39GgOmk6lTZ010t:-KQUKNfARXVXbAG8aWT-:b53yryj) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L271)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Also fix this path.
</details>
Problem solved

---

*[example/assets/normaltexture.jpg, line  \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRtonFcBjFeAOESut1:-KQUKmdXBaNv1Z2o6SoV:b6a0st0) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/assets/normaltexture.jpg#L)):*
<details><summary><i>Previously, ataber (Andrew Taber) wrote…</i></summary>
> Unfortunately I have no idea what the license is, and a google reverse image search only shows other people using the texture on Pinterest, github projects, etc, no original source. Is there a standard place to find free textures like this?
</details>
Go to opengameart and look. For example here:
http://opengameart.org/content/50-free-textures-4-normalmaps

You can also make your own with a noise library like libnoise if you're feeling really ambiguous.
http://libnoise.sourceforge.net/index.html

This program can also be used to procedurally make normal maps:
http://neotextureedit.sourceforge.net/ 

---

*[example/assets/spheretexture.jpg, line  \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRuI8L7FKfSZgDJLjQ:-KQUM4xoP5NgBlltHWyI:b-knjsts) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/assets/spheretexture.jpg#L)):*
<details><summary><i>Previously, ataber (Andrew Taber) wrote…</i></summary>
> Same as the other texture.
</details>
Just search for matcaps on the internet. Like here are some free ones
https://bensimonds.com/2010/07/30/matcap-generator/

or make your own. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/287)*

**Comment by Erkaman** [comment]
Also it would be great if you in the README could write down the exact source of those textures. 
You put it here:
https://github.com/regl-project/regl#license

**Comment by Erkaman** [comment]
Alright, this should be good enough to merge in now. 👍 

But we can't do that until @mikolalysenko fixes CircleCI. So we'll just have to wait a bit.

**Comment by Erkaman** [comment]
Reviewed 1 of 2 files at r8, 2 of 3 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/287)*

**Comment by Erkaman** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[package.json, line 49 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRutj9rLB9chvRFjdH:-KQW_MRgk8-cF3orhxtf:b32w0x) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/package.json#L49)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Alright, problem solved then. 
</details>
OK

---

*[example/metaball.js, line 169 \[r2\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KPu5bed8YCQWJkebDjm-r2-169:-KQW_RsirpHEkEo_hpyZ:b32w0x) ([raw file](https://github.com/regl-project/regl/blob/adfd9128bb872703a9e94d9919f9100f7832d115/example/metaball.js#L169)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Problem solved
</details>
OK

---

*[example/metaball.js, line 133 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRv5oMGuGwo4shhcab:-KQW_OxSUhfBUJ7I2e1r:b32w0x) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L133)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Alright problem solved then
</details>
OK

---

*[example/metaball.js, line 152 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRyJZovKqQUdzdOKuR:-KQW_Q_dRnTVduafXEPI:b32w0x) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L152)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Explain that you are using a matcap here. And maybe link to this article as well:
> https://www.clicktorelease.com/blog/creating-spherical-environment-mapping-shader
> Because this is not entirely obvious the first time you see it. This would make it easier to read. 
> 
> And for introducing more structure to the shader, maybe you should do the matcap stuff in a function of it's own? And also the triplanar texturing in a function of its own? Because the shader is slightly messy now. It's just one big block of code, basically. 
</details>
OK

---

*[example/metaball.js, line 162 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRvfw8V_JdmdKxs3CX:-KQW_RPEMWmDTl-YgApp:b32w0x) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L162)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Problen solved
</details>
OK

---

*[example/metaball.js, line 260 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRwoHIU2QTUD18hluG:-KQW_SWMjoTb_yHVUzs9:b32w0x) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L260)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Problem solved
</details>
OK

---

*[example/metaball.js, line 263 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRxClTq9j3OTaGjNJP:-KQW_T-mPGeimtdYFHtC:b32w0x) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L263)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Problem solved
</details>
OK

---

*[example/metaball.js, line 271 \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRx39GgOmk6lTZ010t:-KQW_TThxuaXE6hiMVBF:b32w0x) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/metaball.js#L271)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Problem solved
</details>
OK

---

*[example/assets/normaltexture.jpg, line  \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRtonFcBjFeAOESut1:-KQW_U4EYvkXcEPcCYxE:b32w0x) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/assets/normaltexture.jpg#L)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Go to opengameart and look. For example here:
> http://opengameart.org/content/50-free-textures-4-normalmaps
> 
> You can also make your own with a noise library like libnoise if you're feeling really ambiguous.
> http://libnoise.sourceforge.net/index.html
> 
> This program can also be used to procedurally make normal maps:
> http://neotextureedit.sourceforge.net/ 
> 
> 
</details>
OK

---

*[example/assets/spheretexture.jpg, line  \[r7\]](https://reviewable.io:443/reviews/regl-project/regl/287#-KQRuI8L7FKfSZgDJLjQ:-KQW_UtrmDcXYmzIvsH6:b32w0x) ([raw file](https://github.com/regl-project/regl/blob/92702f2747748e21c6c6772760b222612d54f686/example/assets/spheretexture.jpg#L)):*
<details><summary><i>Previously, Erkaman (Eric Arnebäck) wrote…</i></summary>
> Just search for matcaps on the internet. Like here are some free ones
> https://bensimonds.com/2010/07/30/matcap-generator/
> 
> or make your own. 
</details>
OK

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/287)*

**Comment by Erkaman** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/287#-:-KQWaGzVh97niIOOYcRT:bnfp4nl)*

**Comment by mikolalysenko** [comment]
If you merge in the master branch then circle-ci should run and we can merge this.  The CI got screwed up when I moved regl into a separate org.

**Comment by Erkaman** [comment]
Yeah, do that @ataber .

**Comment by ataber** [comment]
K merged. I think Reviewable still wants some attention...

**Comment by Erkaman** [comment]
Reviewed 1 of 1 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/regl-project/regl/287)*

**Comment by Erkaman** [comment]
Merging in. Awesome work 👍 
Gonna go and rebuild the gallery now.

**Comment by Erkaman** [comment]
Also, this resolves issue #226.

**Pull request merged**